fmralign/hyperalignment/regions_alignment.py

```diff
@@ -0,0 +1,214 @@
+        ----------
+        X : list of ndarray of shape (n_samples, n_voxels)
+            The brain images for one subject.
+        searchlights : list of searchlights
+            The searchlight indices.
+        dists : list of distances
+
+        radius : int
+            The radius of the searchlight (in millimeters)
+
+        Returns
+        -------
+        Xhat : list of ndarray of shape (n_samples, n_voxels)
+            The denoised estimations B_1, ... B_p for each subject.
+        """
+
+        self.FUNC = "RegionAlignment"
+        if dists is not None and radius is not None:
+            self.FUNC = "SearchlightAlignment"
+
+        if self.verbose:
+            print(f"[{self.FUNC}] Shape of input data: ", X.shape)
+
+        try:
+            self.Xhat = np.load(self.path + "/train_data_denoised.npy")
+            if self.verbose:
+                print(f"[{self.FUNC}] Loaded denoised data from cache")
+            return self.Xhat
+        except:  # noqa E722
+            if self.verbose:
+                print(f"[{self.FUNC}] No cache found, computing denoised data")
+
+        self.n_s, self.n_t, self.n_v = X.shape
+        self.regions = regions
+        self.FUNC = "ParcelAlignment"
+
+        if weights is None and dists is not None and radius is not None:
+            self.distances = dists
+            self.radius = radius
+            self.FUNC = "SearchlightAlignment"
+
+        # Compute global template M (sl_template)
+        if self.verbose:
+            print(f"[{self.FUNC}]Computing global template M ...")
+
+        try:
+            sl_template = np.load(self.path + ("/train_data_template.npy"))
+            if self.verbose:
+                print("Loaded template from cache")
+        except:  # noqa E722
+            if self.verbose:
+                print(f"[{self.FUNC}] No cache found, computing template")
+            if dists is None or radius is None:
+                self.weights = None
+            else:
+                self.weights = searchlight_weights(
+                    searchlights=regions, dists=dists, radius=radius
+                )
+            sl_template = template(
+                X,
+                regions=regions,
+                n_jobs=self.n_jobs,
+                template_kind=self.template_kind,
+                verbose=self.verbose,
+                weights=self.weights,
+            )
+
+        if self.cache:
+            np.save(self.path + ("/train_data_template.npy"), sl_template)
+            if self.verbose:
+                print(f"[{self.FUNC}] Saved template to cache")
+
+        self.Xhat = Parallel(n_jobs=self.n_jobs)(
+            delayed(self.compute_linear_transformation)(X[i], sl_template, i)
+            for i in range(self.n_s)
+        )
+
+        if id is None:
+            id = np.random.randint(0, 1000000)
+
+        if self.cache:
+            np.save(self.path + ("/train_data_denoised.npy"), self.Xhat)
+
+        return np.array(self.Xhat)
```

Comment:
You don't return the template ?

Reply:
This is the denoising class, so I need to return the denoised signal directly